sdk/eventhub/event-hubs/src/eventHubBufferedProducerClient.ts

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AmqpAnnotatedMessage } from "@azure/core-amqp";
+import { NamedKeyCredential, SASCredential, TokenCredential } from "@azure/core-auth";
+import { EventData, OperationOptions } from ".";
+import { ConnectionContext, createConnectionContext } from "./connectionContext";
+import { EventHubProperties, PartitionProperties } from "./managementClient";
+import {
+  EventHubClientOptions,
+  GetEventHubPropertiesOptions,
+  GetPartitionIdsOptions,
+  GetPartitionPropertiesOptions,
+  SendBatchOptions
+} from "./models/public";
+import { isCredential } from "./util/typeGuards";
+
+/**
+ * Contains the events that were successfully sent to the Event Hub,
+ * and the partition they were assigned to.
+ */
+export interface OnSendEventsSuccessContext {
+  /**
+   * The partition each event was assigned.
+   */
+  partitionId: string;
+  /**
+   * The array of {@link EventData} and/or `AmqpAnnotatedMessage` that were successfully sent to the Event Hub.
+   */
+  events: Array<EventData | AmqpAnnotatedMessage>;
+}
+
+/**
+ * Contains the events that were not successfully sent to the Event Hub,
+ * the partition they were assigned to, and the error that was encountered while sending.
+ */
+export interface OnSendEventsErrorContext {
+  /**
+   * The partition each event was assigned.
+   */
+  partitionId: string;
+  /**
+   * The array of {@link EventData} and/or `AmqpAnnotatedMessage` that were not successfully sent to the Event Hub.
+   */
+  events: Array<EventData | AmqpAnnotatedMessage>;
+  /**
+   * The error that occurred when sending the associated events to the Event Hub.
+   */
+  error: Error;
+}
+
+/**
+ * Describes the options that can be provided while creating the `EventHubBufferedProducerClient`.
+ */
+export interface EventHubBufferedProducerClientOptions extends EventHubClientOptions {
+  /**
+   * The maximum number of events to buffer.
+   */
+  maxBufferedEventCount?: number;
+  /**
+   * The maximum amount of time to wait before sending a batch of messages to the Event Hub.
+   */
+  maxWaitTimeInMs?: number;
+  /**
+   * The handler to call once a batch has successfully published.
+   */
+  onSendEventsSuccessHandler?: (ctx: OnSendEventsSuccessContext) => Promise<void>;
+  /**
+   * The handler to call when a batch fails to publish.
+   */
+  onSendEventsErrorHandler: (ctx: OnSendEventsErrorContext) => Promise<void>;
+}
+
+/**
+ * Options to configure the `close` method on the `EventHubBufferedProducerClient`.
+ */
+export interface BufferedCloseOptions extends OperationOptions {
+  /**
+   * When `true`, all buffered events that are pending should be sent before closing.
+   * When `false`, abandon all buffered events and close immediately.
+   * Defaults to `false`.
+   */
+  flush?: boolean;
```

Comment:
It seems like the default should be 'true', not 'false' since that's the safest way to operate.

Any reason to prefer abandon?

Reply:
(and if the default flips maybe you could just change the field to 'abandon', so the default-ish value lines up with what you want)

Reply:
And just to check, I actually thought 'flush' was the default in .NET but that could have changed (CC: @jsquire)

Reply:
Correct;  the intent, based on board feedback, is to prefer `flush` and require opt-in for abandonment.

Reply:
If a similar thing will be added to service-bus, "abandon" wouldn't be a good pick since it is already taken.

Reply:
Ah, I thought I copied what .NET was doing here for flush. @jsquire, do you still call it flush and default to true if the user doesn't specify it then?

Reply:
We do.   _([src](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs#L809))_